tests/tt_metal/tt_metal/unit_tests_fast_dispatch/command_queue/test_write_host.cpp

```diff
@@ -0,0 +1,230 @@
+    uint32_t dev_hugepage_start_diff_uint = (dev_hugepage_start - dev_hugepage_base) / sizeof(uint32_t);
+    uint32_t hugepage_buffer_size_g_uint = hugepage_buffer_size_g / sizeof(uint32_t);
+    int fail_count = 0;
+
+    for (int i = 0; i < num_words; ++i) {
+        uint32_t hugepage_idx = (dev_hugepage_start_diff_uint + i) % hugepage_buffer_size_g_uint;
+        if (results[hugepage_idx] != dev_data[i]) {
+            if (!failed) {
+                tt::log_fatal("Data mismatch");
+                fprintf(stderr, "First 20 failures for each core: [idx] expected->read\n");
+            }
+            if (fail_count == 0) {
+                fprintf(stderr, "Failures reading hugepage\n");
+            }
+
+            fprintf(stderr, "  [%02d] 0x%08x->0x%08x\n", i, (unsigned int)dev_data[i], (unsigned int)results[hugepage_idx]);
+
+            failed = true;
+            fail_count++;
+            if (fail_count > 20) {
+                break;
+            }
+        }
+    }
+
+    return !failed;
+}
+
+namespace local_test_functions {
+
+bool test_write_host(Device *device, uint32_t data_size, std::pair<uint32_t, uint32_t> write_ptr_start = default_ptrs, std::pair<uint32_t, uint32_t> read_ptr_start = default_ptrs, std::optional<std::pair<uint32_t, uint32_t>> read_ptr_update = std::nullopt) {
+    CoreCoord spoof_prefetch_core = {0, 0};
+    CoreCoord dispatch_core = {4, 0};
+    CoreCoord phys_spoof_prefetch_core = device->worker_core_from_logical_core(spoof_prefetch_core);
+    CoreCoord phys_dispatch_core = device->worker_core_from_logical_core(dispatch_core);
+
+    tt::tt_metal::Program program = tt::tt_metal::CreateProgram();
+
+    uint32_t dispatch_buffer_size_blocks_g = 4;
+
+    uint32_t total_size = data_size + sizeof(CQDispatchCmd);
+
+    uint32_t prefetcher_iterations_g = tt::div_up(total_size, dispatch_buffer_page_size_g);
+
+    // Add one page for terminate cmd
+    uint32_t dispatch_buffer_pages = prefetcher_iterations_g + 1;
+    // Make full blocks
+    dispatch_buffer_pages = tt::round_up(dispatch_buffer_pages, dispatch_buffer_size_blocks_g);
+
+    // NOTE: this test hijacks hugepage
+    // Start after first page since ptrs are at the start of hugepage
+
+    void *host_hugepage_base;
+    chip_id_t mmio_device_id = tt::Cluster::instance().get_associated_mmio_device(device->id());
+    uint16_t channel = tt::Cluster::instance().get_assigned_channel_for_device(device->id());
+    host_hugepage_base = (void *)tt::Cluster::instance().host_dma_address(0, mmio_device_id, channel);
+    host_hugepage_base = (void *)((uint8_t *)host_hugepage_base + dev_hugepage_base);
+
+    uint32_t l1_buf_base = L1_UNRESERVED_BASE;
+    uint32_t log_dispatch_buffer_page_size_g = 12;
+    uint32_t dispatch_buffer_size_g = dispatch_buffer_pages * dispatch_buffer_page_size_g;
+
+    TT_FATAL(l1_buf_base + dispatch_buffer_size_g <= device->l1_size_per_core(), "Does not fit in L1");
+
+    std::vector<uint32_t> dispatch_cmds;
```

Comment:
is a shame to duplicate all of this.  maybe fine for now.  hmm.  ideally a header library that the other test infra can share?